### ISSUE TITLE
How to extract the complete data.frame from plot data.

### ISSUE THREAD

**Issue by blavetn** [multimeric/TidyMultiqc#1]
Dear Mr Milton

I have been trying to extract plot data from multiqc.
I couldn't understand how to extract at once all the values for x and y, so I did the following:

```R
gene_cov <- as.data.table(load_multiqc(
    paste0(MQC_DATA, "multiqc_data.json"),
    sections = "plots",
    plot_opts = list(picard_rna_coverage = list(
      extractor = extract_xy,
      summary = list(
        `0` = purrr::partial(summary_extract_df, row_select = x == 0),
        `1` = purrr::partial(summary_extract_df, row_select = x == 1),
        `2` = purrr::partial(summary_extract_df, row_select = x == 2),
        `3` = purrr::partial(summary_extract_df, row_select = x == 3),
        `4` = purrr::partial(summary_extract_df, row_select = x == 4),
        `5` = purrr::partial(summary_extract_df, row_select = x == 5),
        `6` = purrr::partial(summary_extract_df, row_select = x == 6),
        `7` = purrr::partial(summary_extract_df, row_select = x == 7),
        `8` = purrr::partial(summary_extract_df, row_select = x == 8),
        `9` = purrr::partial(summary_extract_df, row_select = x == 9),
        `10` = purrr::partial(summary_extract_df, row_select = x == 10)
      )
    ))
  ))
```

At this point gene_cov look like this:

```
metadata.sample_id	plot.picard_rna_coverage.0	plot.picard_rna_coverage.1	plot.picard_rna_coverage.2	plot.picard_rna_coverage.3 ...
sample1	0.307203	0.352601	0.428728	0.506128	...
sample2	0.337342	0.397810	0.482380	0.562006	...
sample3	0.315192	0.366014	0.437372	0.503733	...
sample4	0.326004	0.382353	0.466560	0.544932	...
... 
```

Then I did this :
```R
names(gene_cov)[-1] = gsub("plot.picard_rna_coverage","x",names(gene_cov)[-1])
setnames(gene_cov, "metadata.sample_id","Sample")

gene_cov <- melt(gene_cov, id.vars = "Sample", variable.name = "x", value.name = "y")
gene_cov <- gene_cov[, x := as.integer(gsub("x\\.","",x))]
```
At this point gene_cov look like this:
```
Sample  x        y
sample1	0	0.307203
sample2	0	0.337342
sample3	0	0.315192
sample4	0	0.326004
sample1	1	0.352601
sample2	1	0.397810
sample3	1	0.366014
sample4	1	0.382353
sample1	2	0.428728
sample2	2	0.482380
sample3	2	0.437372
sample4	2	0.466560
...
```

Please could you tell me what command to use to extract at once all the data from a multiqc plot that I don't have to write the list of all the coordinate ?

Best regards

Nicolas Blavet

**Comment by multimeric** [comment]
Hi Nicolas,

Could you also please post a sample JSON file for this plot type, so I can see what you are working with? It can be the `multiqc_data.json` that you are using, or a different data set that has the same plot type if you aren't able to share your file.

**Comment by blavetn** [comment]
I am sending you an example to your email.

**Comment by multimeric** [comment]
Hi, thanks for the data. In future you should be able to attach files directly to the GitHub issue.

So the main thing to remember here is that each of your summary functions takes a data frame (the entire plot data for that sample) and returns a single value (the summary statistic). What you actually want here is that entire dataframe, without any summary applied to it. Rather than pulling out each data point as you have done and then reconstructing the original data frame, you can instead just nest the whole data frame into your summary data frame by putting it in a list. For example:

```R
library(dplyr)
library(tidyr)
library(TidyMultiqc)

df = load_multiqc(
  "multiqc_data.json",
  sections = "plots",
  plot_opts = list(picard_rna_coverage = list(
    extractor = extract_xy,
    summary = list(
      df=list
    )
  ))
)
```
This will give you:
```
# A tibble: 6 x 2
  metadata.sample_id         plot.picard_rna_coverage.df
  <chr>                      <list>                     
1 Trop_2_BirA_DOX_minus_rep1 <tibble[,2] [101 × 2]>     
2 Trop_2_BirA_DOX_minus_rep2 <tibble[,2] [101 × 2]>     
3 Trop_2_BirA_DOX_minus_rep3 <tibble[,2] [101 × 2]>     
4 Trop_2_BirA_DOX_plus_rep1  <tibble[,2] [101 × 2]>     
5 Trop_2_BirA_DOX_plus_rep2  <tibble[,2] [101 × 2]>     
6 Trop_2_BirA_DOX_plus_rep3  <tibble[,2] [101 × 2]>   
```
In other words, we now have an x, y data frame for each sample. You can then unnest that data frame, which I assume is what you want:
```R
df %>% unnest(plot.picard_rna_coverage.df)
```
```
# A tibble: 606 x 3
   metadata.sample_id             x     y
   <chr>                      <int> <dbl>
 1 Trop_2_BirA_DOX_minus_rep1     0 0.209
 2 Trop_2_BirA_DOX_minus_rep1     1 0.232
 3 Trop_2_BirA_DOX_minus_rep1     2 0.271
 4 Trop_2_BirA_DOX_minus_rep1     3 0.315
 5 Trop_2_BirA_DOX_minus_rep1     4 0.365
 6 Trop_2_BirA_DOX_minus_rep1     5 0.417
 7 Trop_2_BirA_DOX_minus_rep1     6 0.453
 8 Trop_2_BirA_DOX_minus_rep1     7 0.494
 9 Trop_2_BirA_DOX_minus_rep1     8 0.542
10 Trop_2_BirA_DOX_minus_rep1     9 0.588
# … with 596 more rows
```

**Comment by blavetn** [comment]
Great ! That exactly what I needed.
Thank you

**Comment by multimeric** [comment]
No worries! I should really add a section about this in the manual, too.

**Comment by multimeric** [comment]
This is now the default behaviour of `TidyMultiqc`, so you no longer need to (and *can* no longer) use this workaround.
